juju/model.py

```diff
@@ -541,7 +532,7 @@ async def resolve(self, url, architecture,
             raise JujuError('revision and channel are mutually exclusive when deploying a bundle. Please choose one.')
 
         if app_name is None:
-            app_name = url.name
+            app_name = charm_url.name
```

Comment:
Looks like the `charm_url` is returned as a string and blowing up here, and needs to be fixed before further review. See https://github.com/juju/python-libjuju/actions/runs/6642890817/job/18048743466?pr=974#step:6:1553

Reply:
Good spot. Thanks, this is a fairly easy fix, _resolve_charm is only called a few times, and when we do we don't assume it's a string mostly

Reply:
So I've decided to change it to return a charm url